packages/remix-dev/compiler/compileBrowser.ts

```diff
@@ -123,29 +152,108 @@ export const createBrowserCompiler = (
+            incremental: true,
+          })
+        : appCompiler.rebuild());
+
+      invariant(
+        appCompiler.metafile,
+        "Expected app compiler metafile to be defined"
+      );
+    };
+
+    let cssBuildTask = async () => {
+      if (!remixConfig.future.unstable_cssModules) {
+        return;
+      }
+
+      // The types aren't great when combining write: false and incremental: true
+      //  so we need to assert that it's an incremental build
+      cssCompiler = (await (!cssCompiler
+        ? esbuild.build({
+            ...createEsbuildConfig("css", remixConfig, options),
+            metafile: true,
+            incremental: true,
+            write: false,
+          })
+        : cssCompiler.rebuild())) as esbuild.BuildIncremental;
+
+      invariant(
+        cssCompiler.metafile,
+        "Expected CSS compiler metafile to be defined"
+      );
+
+      let cssBundlePath: string | undefined;
+      let outputFiles = cssCompiler.outputFiles || [];
+
+      await Promise.all(
+        outputFiles.map(async (outputFile) => {
+          let outputPath = outputFile.path;
+
+          // Only write bundled CSS and source map to disk
+          if (
+            path.dirname(outputPath) === remixConfig.assetsBuildDirectory &&
+            path.basename(outputPath).startsWith("css-bundle")
+          ) {
+            if (outputPath.endsWith(".css")) {
+              // Grab the CSS bundle path so we can use it to generate the manifest
+              cssBundlePath = outputPath;
+              await fse.ensureDir(path.dirname(outputPath));
+
+              let contents =
+                options.mode === "production"
+                  ? await postcss([
+                      // We need to discard duplicate rules since "composes"
+                      // in CSS Modules can result in duplicate styles
+                      postcssDiscardDuplicates(),
+                    ])
+                      .process(
+                        Buffer.from(outputFile.contents).toString("utf-8")
+                      )
+                      .then((result) => result.css)
+                  : outputFile.contents;
+
+              return await fse.writeFile(outputPath, contents);
+            }
+
+            if (outputPath.endsWith(".css.map")) {
+              await fse.ensureDir(path.dirname(outputPath));
+              return await fse.writeFile(outputPath, outputFile.contents);
+            }
+          }
+
+          return null;
+        })
+      );
+
+      return cssBundlePath;
+    };
+
+    let [cssBundlePath] = await Promise.all([cssBuildTask(), appBuildTask()]);
+
+    let manifest = await createAssetsManifest({
+      config: remixConfig,
+      metafile: appCompiler.metafile!,
+      cssBundlePath,
+    });
     manifestChannel.write(manifest);
     await writeAssetsManifest(remixConfig, manifest);
   };
+
   return {
     compile,
-    dispose: () => compiler?.rebuild.dispose(),
+    dispose: () => {
```

Comment:
Can't remember if these are synchronous, so ... just checking that there aren't any potential race conditions here.

Reply:
This is a good question, I wasn't sure myself until double checking. It's sync, so we're good 👍